src/TupleTools.jl

```diff
@@ -327,4 +324,13 @@ Inverse permutation of a permutation `p`.
 """
 invperm(p::Tuple{Vararg{Int}}) = sortperm(p)
 
+
+@generated function diff(v::NTuple{N, T}) where {N, T}
```

Comment:
the point of TupleTools.jl is not too use `@generated` functions but rather recursive definitions, i.e.
```julia
diff(v::Tuple{}) = () # similar to diff([])
diff(v::Tuple{Any}) = ()
diff(v::Tuple) = (v[2]-v[1], diff(Base.tail(v))...)
```

Reply:
Why recursive is preferred? You recursive version is 140x slower than mine

```julia
julia> @generated function gen_diff(v::NTuple{N, T}) where {N, T}
           ex = Expr(:tuple)
           for k in 1:N-1
               push!(ex.args, :(v[$(k+1)] - v[$k]))
           end
           return ex
       end
gen_diff (generic function with 1 method)

julia> @benchmark gen_diff($(Tuple(1:20)))
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     0.036 ns (0.00% GC)
  median time:      0.040 ns (0.00% GC)
  mean time:        0.040 ns (0.00% GC)
  maximum time:     0.048 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
```

```julia
julia> @benchmark diff($(Tuple(1:20)))
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     5.314 ns (0.00% GC)
  median time:      5.685 ns (0.00% GC)
  mean time:        7.669 ns (0.00% GC)
  maximum time:     74.435 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
```

Isn't the whole point of using `Tuple` instead of a normal `Vector` is that `Tuple` provides more info during compile time and can make it much faster? And this package is called `TupleTools` not `RecursiveTupleTools`...

Reply:
Normally one shouldn't trust timings below the nanosecond limit. Up to comment lines, the `@code_llvm` output of the recursive and the `@generated` definition is almost identical, up to a slightly different order in the first few comments:
https://gist.github.com/Jutho/93be2e4ff71944bc7a5e33b3052d82db

Not sure though why `@benchmark` is reporting such a big difference.

Reply:
neither `code_llvm` nor `code_native` are same on my side, what's your Julia version? This is mine

```julia
julia> versioninfo()
Julia Version 1.0.1
Commit 0d713926f8 (2018-09-29 19:05 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.0.0)
  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libimf
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
```

Reply:
Remove all the comment lines (starting with semicolumn `;`) and put them next to each other. They are exactly as long and contain the same statements, the only difference is that the order of the first half of the lines is somewhat different.

Really these things do the same: consider this:
```julia
julia> @generated function gen_diff(v::NTuple{N, T}) where {N, T}
                  ex = Expr(:tuple)
                  for k in 1:N-1
                      push!(ex.args, :(v[$(k+1)] - v[$k]))
                  end
                  return ex
              end
julia> rec_diff(v::Tuple) = (v[2]-v[1], rec_diff(Base.tail(v))...)
julia> rec_diff(v::Tuple{Any}) = ()
julia> rec_diff(v::Tuple{}) = () # similar to diff([])
julia> function f(g)
           t = ntuple(n->0, Val(20))
           for i = 1:1000
               s = ntuple(n->i+n, Val(21))
               t = t .+ g(s)
           end
           return t
       end
julia> using BenchmarkTools
julia> @benchmark f($gen_diff)
BenchmarkTools.Trial: 
  memory estimate:  176 bytes
  allocs estimate:  1
  --------------
  minimum time:     23.187 ns (0.00% GC)
  median time:      24.459 ns (0.00% GC)
  mean time:        31.879 ns (17.86% GC)
  maximum time:     35.018 μs (99.88% GC)
  --------------
  samples:          10000
  evals/sample:     994

julia> @benchmark f($rec_diff)
BenchmarkTools.Trial: 
  memory estimate:  176 bytes
  allocs estimate:  1
  --------------
  minimum time:     22.204 ns (0.00% GC)
  median time:      22.806 ns (0.00% GC)
  mean time:        30.970 ns (18.61% GC)
  maximum time:     36.592 μs (99.88% GC)
  --------------
  samples:          10000
  evals/sample:     994
```

Here `rec_diff` and `gen_diff` are run a thousand times, together with other statements (i.e. `.+` and tuple creation). So a single run is not taking 5 ns, and not even 0.036 ns. But most importantly, they take exactly as long.

Reply:
True. I can reproduce your results now. Thanks.